chaincode/src/launchpad/createSale.ts

```diff
@@ -0,0 +1,145 @@
+import { createTokenClass } from "@gala-chain/chaincode";
+import { BigNumber } from "bignumber.js";
+
+import { buyWithNative } from "./buyWithNative";
+
+BigNumber.config({
+  ROUNDING_MODE: BigNumber.ROUND_UP
+});
+
+/**
+ * Creates a new token sale (LaunchPad) in the GalaChain environment.
+ *
+ * This function validates input parameters, creates a token class, mints an initial supply
+ * of tokens, and initializes the LaunchPad sale object. If a pre-buy amount is specified,
+ * it simulates a token purchase using native tokens.
+ *
+ * @param ctx - The context object providing access to the GalaChain environment.
+ * @param launchPadDetails - An object containing details for the token sale, including:
+ *   - `tokenName`: The name of the token being created.
+ *   - `tokenSymbol`: The symbol for the token.
+ *   - `tokenDescription`: A description of the token.
+ *   - `tokenImage`: An optional image URL for the token.
+ *   - `preBuyAmount`: The amount of native tokens to use for a pre-buy (optional).
+ *
+ * @returns A promise that resolves to a `CreateSaleResponse` object containing details about
+ *          the created sale, including the vault address and token collection name.
+ *
+ * @throws DefaultError if:
+ *   - Required fields (`tokenName`, `tokenSymbol`, `tokenDescription`) are missing or empty.
+ *   - The token class creation fails (no response key returned).
+ */
+export async function createSale(
+  ctx: GalaChainContext,
+  launchPadDetails: CreateTokenSaleDTO
+): Promise<CreateSaleResponse> {
+  // Validate input parameters
+  if (!launchPadDetails.websiteUrl && !launchPadDetails.telegramUrl && !launchPadDetails.twitterUrl) {
+    throw new DefaultError("Token sale creation requires atleast one social link.");
+  }
+
+  launchPadDetails.tokenSymbol = launchPadDetails.tokenSymbol.toUpperCase();
+
+  // Define the token class key
+  const tokenInstanceKey = new TokenInstanceKey();
+  tokenInstanceKey.collection = "Token";
+  tokenInstanceKey.category = "Unit";
+  tokenInstanceKey.type = `${launchPadDetails.tokenSymbol}`;
+  tokenInstanceKey.additionalKey = `${ctx.callingUser.replace(/\|/, ":")}`;
+  tokenInstanceKey.instance = new BigNumber(0);
+
+  // Validate uniqueness of sale and token
+  const vaultAddress = `service|${tokenInstanceKey.getTokenClassKey().toStringKey()}$launchpad`;
+  const key = ctx.stub.createCompositeKey(LaunchPadSale.INDEX_KEY, [vaultAddress]);
+  const sale = await getObjectByKey(ctx, LaunchPadSale, key).catch(() => undefined);
+  if (sale) {
+    throw new DefaultError("This token and a sale associtated with it already exists");
+  }
+
+  // Call createTokenClass
+  const responseKey = await createTokenClass(ctx, {
+    network: "GC",
+    tokenClass: tokenInstanceKey.getTokenClassKey(),
+    isNonFungible: false,
+    decimals: 18,
+    name: launchPadDetails.tokenName,
+    symbol: launchPadDetails.tokenSymbol,
+    description: launchPadDetails.tokenDescription,
+    rarity: "abc",
```

Comment:
Similar to the above comment, I question the approach of hard-coding these values. 

Why not allow the token creator to specify a value here?

Or at the very least, choose a value that makes more sense for production code than "abc"?

```suggestion
    rarity: launchPadDetails.tokenRarity,
```